### PR TITLE
feat(batch): prepare for union operator

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR implements the basic batch union operator.
- It is a preparation for the later implementation like: merge index and set operator like union all.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#2912  #4339

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_union.rs [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+use risingwave_common::error::Result;
+
+use super::{ColPrunable, PlanBase, PlanRef, PredicatePushdown, ToBatch, ToStream};
+use crate::optimizer::plan_node::{BatchHashAgg, BatchUnion, LogicalAgg, PlanTreeNode};
+use crate::optimizer::property::FunctionalDependencySet;
+use crate::utils::{ColIndexMapping, Condition};
+
+/// `LogicalUnion`
+///  returns the union of the rows of its inputs
+///  if all is false means it need to eliminate duplicates.
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/logical_union.rs:27]:
```suggestion
/// `LogicalUnion` returns the union of the rows of its inputs.
/// If `all` is false, it needs to eliminate duplicates.
```

File: src/frontend/src/optimizer/plan_node/logical_union.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        self.clone_with_inputs(&new_inputs)
+    }
+}
+
+impl ToBatch for LogicalUnion {
+    fn to_batch(&self) -> Result<PlanRef> {
+        let new_inputs: Result<Vec<_>> =
+            self.inputs().iter().map(|input| input.to_batch()).collect();
+        let new_logical = Self::new(true, new_inputs?);
+        // convert union to union all + agg
```
Comment (xxchan) [inline src/frontend/src/optimizer/plan_node/logical_union.rs:110]: Will the `BatchUnion` executor support dedup in the future? 🤔

**Review by xxchan — APPROVED** [review]
Generally LGTM

File: src/frontend/src/optimizer/plan_node/logical_union.rs (hunk 2 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/logical_union.rs:110]: I think `HashAggExecutor` is good enough, we don't need to implement the dedup logic again. BTW, union all is some kind of exchange. I think we even can use a special exchange executor to replace `UnionExecutor`.

**Pull request merged**